oauth2client/transport/httplib2.py

```diff
@@ -62,98 +62,56 @@ def inject_credentials(credentials, http, refresh_status_codes):
     def new_request(uri, method='GET', body=None, headers=None,
                     redirections=httplib2.DEFAULT_MAX_REDIRECTS,
                     connection_type=None):
-        if not credentials.access_token:
-            _LOGGER.info('Attempting refresh to obtain '
-                         'initial access_token')
-            credentials._refresh(orig_request_method)
 
         # Clone and modify the request headers to add the appropriate
         # Authorization header.
         headers = _helpers.initialize_headers(headers)
-        credentials.apply(headers)
         _helpers.apply_user_agent(headers, credentials.user_agent)
 
+        credentials.before_request(
+            orig_request_method, uri, headers)
+        headers = _helpers.clean_headers(headers)
+
         body_stream_position = None
         # Check if the body is a file-like stream.
         if all(getattr(body, stream_prop, None) for stream_prop in
                _STREAM_PROPERTIES):
             body_stream_position = body.tell()
 
-        resp, content = request(orig_request_method, uri, method, body,
-                                _helpers.clean_headers(headers),
-                                redirections, connection_type)
+        resp, content = request(
+            orig_request_method, uri, method, body,
+            headers, redirections, connection_type)
 
         # A stored token may expire between the time it is retrieved and
         # the time the request is made, so we may need to try twice.
         max_refresh_attempts = 2
         for refresh_attempt in range(max_refresh_attempts):
             if resp.status not in refresh_status_codes:
                 break
-            _LOGGER.info('Refreshing due to a %s (attempt %s/%s)',
-                         resp.status, refresh_attempt + 1,
-                         max_refresh_attempts)
-            credentials._refresh(orig_request_method)
-            credentials.apply(headers)
-            if body_stream_position is not None:
-                body.seek(body_stream_position)
 
-            resp, content = request(orig_request_method, uri, method, body,
-                                    _helpers.clean_headers(headers),
-                                    redirections, connection_type)
+            _LOGGER.info(
+                'Refreshing due to a %s (attempt %s/%s)',
+                resp.status, refresh_attempt + 1, max_refresh_attempts)
 
-        return resp, content
+            credentials.refresh(orig_request_method)
 
-    # Replace the request method with our own closure.
-    http.request = new_request
+            # Remove the existing Authorization header, as the credentials
+            # may set the header as u'Authorization' but clean_headers set it
+            # as b'Authorization'.
+            headers.pop(b'Authorization', None)
```

Comment:
So here's what happens:

1. First request
  1. `credentials.apply` sets `headers[u'Authorization']`
  2. `clean_headers` renames `headers[u'Authorization']` to `headers[b'Authorization']`
  3. Receive 401 response, so retry the request. *The current headers are re-used*.
2. Retry request.
  1. `credentials.refresh` is called.
  2. `credentials.apply` sets `headers[u'Authorization']`.
  3. At this point `headers` is `{b'Authorization': 'old token', u'Authorization': 'new token'}`
  4. clean_headers may or may not use the second one or the first one when sanitizes because of non-deterministic dictionary ordering.

Imagine how long it too me to figure this out. :)